scripts/hallucinate_fragalysis.py

```diff
@@ -0,0 +1,589 @@
+            sort_key = "chemgauss4_score"
+        best_str_dict[(compound_id, dimer)] = g.loc[
+            sort_fn(g[sort_key]), f"MERS_structure"
+        ]
+
+    ## Set up SDF file that will hold all ligands
+    all_ligs_monomer_ofs = oechem.oemolostream()
+    all_ligs_monomer_ofs.SetFlavor(
+        oechem.OEFormat_SDF, oechem.OEOFlavor_SDF_Default
+    )
+    all_ligs_dimer_ofs = oechem.oemolostream()
+    all_ligs_dimer_ofs.SetFlavor(
+        oechem.OEFormat_SDF, oechem.OEOFlavor_SDF_Default
+    )
+    # arranged this way so dimer=True => "dimer"
+    dimers_strings = ["monomer", "dimer"]
+    all_ofs = [all_ligs_monomer_ofs, all_ligs_dimer_ofs]
+    for d, ofs in zip(dimers_strings, all_ofs):
+        os.makedirs(f"{out_dir}/{d}", exist_ok=True)
+        ofs.open(f"{out_dir}/{d}/combined.sdf")
+
+    ## Loop through dict and parse input files into output files
+    for (compound_id, dimer), best_str in best_str_dict.items():
+        ## Make sure input exists
+        dimer_s = "dimer" if dimer else "monomer"
+        tmp_in_dir = f"{in_dir}/{compound_id}/{dimer_s}/{best_str}"
+        if not check_output(tmp_in_dir):
+            print(
+                (
+                    f"No results found for {compound_id}/{dimer_s}/{best_str}, "
+                    "skipping"
+                ),
+                flush=True,
+            )
+        ## Create output directory if necessary
+        tmp_out_dir = f"{out_dir}/{dimer_s}/{compound_id}"
+        os.makedirs(tmp_out_dir, exist_ok=True)
+
+        ## Load necessary files
+        du = oechem.OEDesignUnit()
+        oechem.OEReadDesignUnit(f"{tmp_in_dir}/predocked.oedu", du)
+        prot = oechem.OEGraphMol()
+        du.GetProtein(prot)
+        lig = load_openeye_sdf(f"{tmp_in_dir}/docked.sdf")
+
+        ## Set ligand title
+        lig.SetTitle(f"{compound_id}_{best_str}")
+
+        ## First save apo
+        save_openeye_pdb(prot, f"{tmp_out_dir}/{best_str}_apo.pdb")
+
+        ## Combine protein and ligand
+        oechem.OEAddMols(prot, lig)
+
+        # ## Get number of residues per chain
+        # num_res_chain_dict = {}
+        # hv = oechem.OEHierView(prot)
+        # for chain in hv.GetChains():
+        #     max_resid = 0
+        #     for frag in chain.GetFragments():
+        #         frag_max = max(
+        #             [r.GetResidueNumber() for r in frag.GetResidues()]
+        #         )
+        #         if frag_max > max_resid:
+        #             max_resid = frag_max
+        #     num_res_chain_dict[chain.GetChainID()] = max_resid
+        # ## Rename and renumber ligand residue
+        # for a in prot.GetAtoms():
+        #     r = oechem.OEAtomGetResidue(a)
+        #     if r.GetName() == "UNL":
+        #         r.SetHetAtom(True)
+        #         r.SetName("LIG")
+        #         r.SetResidueNumber(num_res_chain_dict[r.GetChainID()] + 100)
+        #     oechem.OEAtomSetResidue(a, r)
```

Comment:
```suggestion
```

Reply:
this is code that's in progress but not quite done yet. I'm planning on opening a PR post-merge to finish, but happy to get rid of the commented code or move it somewhere else for now if you think that's the right thing to do

Reply:
I would say for now to get rid of it, unless you plan on merging the PR into `hallucinate-fragalysis` before we merge this to `main`. In general the idea is to keep commented code out of the `main` branch since we want to use comments for comments. I know things are in flux but I want to try and keep `main` as clean as possible, and we can keep the messy stuff in the branches :smile:

Reply:
made issue here: https://github.com/choderalab/covid-moonshot-ml/issues/50